### PR TITLE
[Exploratory Worker] Add endpoints for launching exploratory db, taking snapshot from replica, and stopping exploratory db

### PR DESCRIPTION
**Pull request by timlee0119**

### CODE REVIEW

File: tmp/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
```
Comment (kush789) [inline tmp/.gitignore:1]: Add to main .gitignore, /tmp/*

File: noisepagecontrol/exploratory_worker/views.py [hunk 2]
```diff
@@ -1,5 +1,70 @@
...
+import subprocess
+import pathlib
+import os
+from threading import Thread
+from django.http import HttpResponse, HttpResponseBadRequest
+from django.views.decorators.csrf import csrf_exempt
+from django.views.decorators.http import require_http_methods
+
+DIR_PATH = pathlib.Path(__file__).resolve().parent
+START_EXPLORATORY_SCRIPT = os.path.join(DIR_PATH, 'scripts/start_exploratory_cluster.sh')
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/views.py]: Add scripts via settings.py

File: noisepagecontrol/logs/.gitignore [hunk 3]
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
```
Comment (kush789) [inline noisepagecontrol/logs/.gitignore:1]: add to main gitignore

File: noisepagecontrol/exploratory_worker/scripts/start_exploratory_cluster.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+##################################  NOTE  ##################################
+#   This script spawns a exploratory Postgres cluster that optionally takes a snapshot
+#   from the production replica cluster.
+#
+#   Script should be run as postgres user;
+#       $ sudo -u postgres scripts/start_exploratory_cluster_snapshot.sh port [-s]
+###############################################################################
+
+port=$1
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/scripts/start_exploratory_cluster.sh]: nit: capitalise all

File: noisepagecontrol/exploratory_worker/scripts/stop_exploratory_cluster.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+POSTGRES_BIN_DIR="/usr/lib/postgresql/14/bin"
+
+# This specifies where the postgres cluster should reside
+#   Ubuntu: /var/lib/postgresql/14
+POSTGRES_ROOT="/var/lib/postgresql/14"
+
+###############################################################################
+
+# TODO Tim: pass in CLUSTER_DIR path
+CLUSTER_DIR="${POSTGRES_ROOT}/exploratory"
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/scripts/stop_exploratory_cluster.sh]: Pending, should query db to find data dir (check get_data_dir.py in primary_executor)

File: noisepagecontrol/exploratory_worker/urls.py [hunk 6]
```diff
@@ -4,4 +4,7 @@
 
 urlpatterns = [
     path("", views.index, name="index"),
+    path("launch_exploratory_cluster", views.launch_exploratory_cluster, name="launch exploratory cluster"),
+    path("collect_training_data", views.collect_training_data, name="collect training data"),
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/urls.py]: Let's remove this for now

File: noisepagecontrol/exploratory_worker/views.py [hunk 7]
```diff
@@ -1,5 +1,70 @@
...
+    thread = Thread(target=start_exploratory_cluster, args=(snapshot,))
+    thread.start()
+
+    return HttpResponse()
+
+
+def start_exploratory_cluster(snapshot):
+    # TODO Tim: probe an available port from 10000 instead of hard-coding
+    exploratory_cluster_port = 10000
+    print(f"starting exploratory Postgres cluster on port {exploratory_cluster_port}...")
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/views.py]: use logger

File: noisepagecontrol/exploratory_worker/views.py [hunk 8]
```diff
@@ -1,5 +1,70 @@
...
+@csrf_exempt
+@require_http_methods(["DELETE"])
+def stop_exploratory_cluster(request):
+    body = json.loads(request.body)
+    port = body['port']
+    
+    print(f"Stopping exploratory Postgres cluster on port {port}...")
+    args = ['sudo', '-u', 'postgres', STOP_EXPLORATORY_SCRIPT, str(port)]
+    res = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    res.wait()
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/views.py]: Does this take time? Should we ack later?

File: noisepagecontrol/exploratory_worker/scripts/start_exploratory_cluster.sh [hunk 9]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/scripts/start_exploratory_cluster.sh]: Move to common scripts folder

File: noisepagecontrol/exploratory_worker/views.py [hunk 10]
```diff
@@ -1,5 +1,70 @@
...
+def launch_exploratory_cluster(request):
+    snapshot = json.loads(request.body)['snapshot']
+    # asynchronously spins up a pg instance
+    thread = Thread(target=start_exploratory_cluster, args=(snapshot,))
+    thread.start()
+
+    return HttpResponse()
+
+
+def start_exploratory_cluster(snapshot):
```
Comment (kush789) [inline noisepagecontrol/exploratory_worker/views.py]: Let's refactor this out into a new service; Check how primary_worker is organised

File: noisepagecontrol/exploratory_worker/scripts/start_exploratory_cluster.sh (hunk 9 above)
Comment (timlee0119) [reply noisepagecontrol/exploratory_worker/scripts/start_exploratory_cluster.sh]: aye sir

File: noisepagecontrol/exploratory_worker/views.py (hunk 8 above)
Comment (timlee0119) [reply noisepagecontrol/exploratory_worker/views.py]: This should be pretty instant so I guess acking synchonously make things easier here?

File: noisepagecontrol/logs/.gitignore (hunk 3 above)
Comment (timlee0119) [reply noisepagecontrol/logs/.gitignore:1]: git doesn't allow an empty folder to be committed so the only way to retain the folder is to either
1. add an empty .gitkeep file under this folder
2. add .gitignore like this
imo the second way is more expressive

File: tmp/.gitignore (hunk 1 above)
Comment (timlee0119) [reply tmp/.gitignore:1]: same

File: .gitignore [hunk 11]
```diff
@@ -5,4 +5,4 @@ erl_crash.dump
 production_cluster
 noisepagecontrol/primary_worker/workload_captures/*
 noisepagecontrol/control_plane/resources/*
-noisepagecontrol/logs/*
```
Comment (kush789) [inline .gitignore:8]: please add this back

File: noisepagecontrol/logs/.gitignore (hunk 3 above)
Comment (kush789) [inline noisepagecontrol/logs/.gitignore:1]: file not needed?

File: scripts/start_exploratory_cluster.sh [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+#       OSX:    /usr/local/bin
+POSTGRES_BIN_DIR="/usr/lib/postgresql/14/bin"
+
+# This specifies where the postgres cluster should reside
+#   Ubuntu: /var/lib/postgresql/14
+POSTGRES_ROOT="/var/lib/postgresql/14"
+
+###############################################################################
+
+EXP_CLUSTER_DIR="$POSTGRES_ROOT/exploratory"  # TODO: parameterize
```
Comment (kush789) [inline scripts/start_exploratory_cluster.sh]: let's change this to `exploratory_${PORT}`? Will allow for multiple spawns

File: tmp/.gitignore (hunk 1 above)
Comment (kush789) [inline tmp/.gitignore:1]: dont need this file?

**Review by kush789 — APPROVED** [review]
lgtm; minor nits

File: .gitignore (hunk 11 above)
Comment (timlee0119) [reply .gitignore:8]: don't have to add this back as the rule is covered in noisepagecontrol/logs/.gitignore

**Pull request merged**